### PR TITLE
Enable activation offloading for Intel XPU backend

### PR DESCRIPTION
**Pull request by zxd1997066**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Enable activation offloading for XPU

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [X] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by fengyuan14** [comment]
Hi, @felipemello1, May I ask your review for the PR? Thanks.

**Comment by felipemello1** [comment]
hey @zxd1997066 and @fengyuan14 . Thanks for the PR! At first glance, it looks good. Were you able to run this with cuda on main and then on this PR on any distributed setting? I would be interested in checking the WandB graphs, to see if there is any regression on memory/tokens per second. To get it, just run for a few steps the config with the flag "metric_logger=torchtune.training.metric_logging.WandBLogger"

I assume that you have also tested on XPU

If everything looks fine, i am happy to merge it.

not a hard requirement for this PR, but i am thinking that we could add a constant somewhere for accelerators_supported, so we dont have to do if cuda or if xpu or if X.

File: torchtune/training/_activation_offloading.py [hunk 1]
```diff
@@ -88,12 +88,17 @@ def __init__(
         self.virtual_memory_safe_pct = (
             60  # we should not exceed this percentage of memory
         )
-
-        self.s0 = torch.cuda.default_stream()  # comp stream
+        # comp stream
+        if torch.cuda.is_available() or torch.xpu.is_available():
```
Comment (majing921201) [inline torchtune/training/_activation_offloading.py]:
```suggestion
        if torch.accelerator.is_available():
```

File: torchtune/training/_activation_offloading.py [hunk 2]
```diff
@@ -147,7 +152,7 @@ def pack_tensor(activation: torch.Tensor) -> int:
             # only offload hefty bois if they're activations on CUDA (our heuristic
             # for that is to check if they're not params or buffers)!
             if (
-                activation.is_cuda
+                activation.is_cuda or activation.is_xpu
```
Comment (majing921201) [inline torchtune/training/_activation_offloading.py]:
```suggestion
                activation.device.type == torch.accelerator.current_accelerator().type
```

**Comment by fengyuan14** [comment]
> hey @zxd1997066 and @fengyuan14 . Thanks for the PR! At first glance, it looks good. Were you able to run this with cuda on main and then on this PR on any distributed setting? I would be interested in checking the WandB graphs, to see if there is any regression on memory/tokens per second. To get it, just run for a few steps the config with the flag "metric_logger=torchtune.training.metric_logging.WandBLogger"
> 
> I assume that you have also tested on XPU
> 
> If everything looks fine, i am happy to merge it.
> 
> not a hard requirement for this PR, but i am thinking that we could add a constant somewhere for accelerators_supported, so we dont have to do if cuda or if xpu or if X.

Thanks for the reply and the remainder. Yes, we have tested the PR on XPU.

Regarding CUDA, we didn't test functionality and performance. For functionality regressions, we thought CI would gate them.
1. Is `distributed setting + activation offloading` covered by unit tests in CI?
2. For memory/performance regression, we can test.

**Comment by zxd1997066** [comment]
hey @felipemello1 Yes, we selected the meta-llama/Meta-Llama-3.1-8B-Instruct lora single device to test the PR on XPU. Below is the WandB runs, and I have set the project to public.
**on XPU:**
enable_activation_offloading=False
```
tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device device=xpu max_steps_per_epoch=50 metric_logger=torchtune.training.metric_logging.WandBLogger
```
WandB runs: https://wandb.ai/eastward-intel/torchtune/runs/nw7e9xiq

enable_activation_offloading=True
```
tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device max_steps_per_epoch=50 enable_activation_offloading=True metric_logger=torchtune.training.metric_logging.WandBLogger
```
WandB runs: https://wandb.ai/eastward-intel/torchtune/runs/exta55l1

Regarding CUDA, we do not have distributed system now, we only tested the single device:
**on main:**
enable_activation_offloading=False
```
tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device max_steps_per_epoch=50 metric_logger=torchtune.training.metric_logging.WandBLogger
```
WandB runs: https://wandb.ai/eastward-intel/torchtune/runs/6b4xyixq

enable_activation_offloading=True
```
tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device max_steps_per_epoch=50 enable_activation_offloading=True metric_logger=torchtune.training.metric_logging.WandBLogger
```
WandB runs: https://wandb.ai/eastward-intel/torchtune/runs/exta55l1

**on this PR**
enable_activation_offloading=False
```
tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device max_steps_per_epoch=50 metric_logger=torchtune.training.metric_logging.WandBLogger
```
WandB runs: https://wandb.ai/eastward-intel/torchtune/runs/2jelkkxh

enable_activation_offloading=True
```
tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device max_steps_per_epoch=50 enable_activation_offloading=True metric_logger=torchtune.training.metric_logging.WandBLogger
```
WandB runs: https://wandb.ai/eastward-intel/torchtune/runs/i5zgo6s1

Is distributed setting + activation offloading could be covered by unit tests in CI ?

**Comment by felipemello1** [comment]
thanks for doing the wandb tests! Could you please just fix the lint? I will merge after it

you can run:
pre-commit install
pre-commit run --all-files

> Is distributed setting + activation offloading could be covered by unit tests in CI ?

We have unit tests, they dont just cover regression :/ https://github.com/pytorch/torchtune/blob/38edb21f4926c4b40b93d9e8b4080f05ddf2da1a/tests/torchtune/training/test_activation_offloading.py#L48
We hope that in the future we can cover memory/TPS, but its so hard because these can be flaky.

**Comment by zxd1997066** [comment]
> thanks for doing the wandb tests! Could you please just fix the lint? I will merge after it
> 
> you can run: pre-commit install pre-commit run --all-files
> 
> > Is distributed setting + activation offloading could be covered by unit tests in CI ?
> 
> We have unit tests, they dont just cover regression :/
> 
> https://github.com/pytorch/torchtune/blob/38edb21f4926c4b40b93d9e8b4080f05ddf2da1a/tests/torchtune/training/test_activation_offloading.py#L48
> 
> 
> We hope that in the future we can cover memory/TPS, but its so hard because these can be flaky.

Thanks, I fixed the lint issue.
